pkg/util/vsphere/watcher/watcher.go

```diff
@@ -0,0 +1,674 @@
+		}()
+
+		if err := w.pc.WaitForUpdatesEx(
+			ctx,
+			&property.WaitOptions{},
+			func(ou []vimtypes.ObjectUpdate) bool {
+				return w.onUpdate(ctx, ou)
+			}); err != nil {
+
+			w.setErr(err)
+		}
+	}()
+
+	return w, nil
+}
+
+const (
+	virtualMachineType               = "VirtualMachine"
+	configPropPath                   = "config"
+	extraConfigPropPath              = configPropPath + ".extraConfig"
+	extraConfigNamespaceNameKey      = "vmservice.namespacedName"
+	extraConfigNamespaceNamePropPath = extraConfigPropPath + `["` + extraConfigNamespaceNameKey + `"]`
+)
+
+type objUpdate struct {
+	kind    vimtypes.ObjectUpdateKind
+	changes []vimtypes.PropertyChange
+}
+
+func (w *Watcher) onUpdate(
+	ctx context.Context,
+	ou []vimtypes.ObjectUpdate) bool {
+
+	logger := logr.FromContextOrDiscard(ctx)
+	logger.V(4).Info("onUpdate", "objectUpdates", ou)
+
+	updates := map[moRef]objUpdate{}
+
+	for i := range ou {
+		oui := ou[i]
+		if oui.Kind != vimtypes.ObjectUpdateKindLeave {
+			if v, ok := updates[oui.Obj]; !ok {
+				updates[oui.Obj] = objUpdate{
+					kind:    oui.Kind,
+					changes: oui.ChangeSet,
+				}
+			} else {
+				v.changes = append(v.changes, oui.ChangeSet...)
+				updates[oui.Obj] = v
+			}
+		}
+	}
+
+	for obj, update := range updates {
+		if err := w.onObject(
+			ctx,
+			obj,
+			update); err != nil {
+
+			w.setErr(err)
+			return true
+		}
+	}
+
+	return false
+}
+
+func (w *Watcher) onObject(
+	ctx context.Context,
+	obj moRef,
+	update objUpdate) error {
+
+	logger := logr.FromContextOrDiscard(ctx).
+		V(4).
+		WithName("onObject").
+		WithValues("obj", obj)
+
+	var (
+		namespace string
+		name      string
+		verified  bool
+	)
+
+	// This update will be skipped if after removing all of the changes for
+	// the ignoredExtraConfigKeys there is nothing left.
+	var ignoredChanges int
+	for i := range update.changes {
+		c := update.changes[i]
+		ignore := false
+		if c.Name == extraConfigPropPath {
+			if aov, ok := c.Val.(vimtypes.ArrayOfOptionValue); ok {
+				ignore, namespace, name = checkExtraConfig(
+					aov, w.ignoredExtraConfigKeys)
+			}
+		}
+		if ignore {
+			ignoredChanges++
+		}
+	}
+	if ignoredChanges == len(update.changes) {
+		return nil
+	}
+
+	if w.lookupNamespacedName != nil {
+		if namespace == "" || name == "" || update.kind == vimtypes.ObjectUpdateKindEnter {
+			namespace, name, verified = w.lookupNamespacedName(ctx, obj)
+		}
+	}
+
+	if update.kind == vimtypes.ObjectUpdateKindEnter && verified {
+		// The behavior of Controller-Runtime to sync all objects upon startup
+		// will cause *existing* VMs to be reconciled. Therefore, do not emit a
+		// result when the object is entering the scope of the watcher and the
+		// corresponding Kubernetes object already exists with a matching
+		// status.uniqueID field.
+		return nil
+	}
+
+	if namespace == "" || name == "" {
+		rep, err := property.DefaultCollector(w.client).RetrieveProperties(
+			ctx,
+			vimtypes.RetrieveProperties{
```

Comment:
You can use `RetrieveOne` (or `Retrieve`) with a `dst` of `*[]types.ObjectContent`. That avoids reflection and having to construct the PropertyFilterSpec:
```go
		var content []types.ObjectContent
		err := property.DefaultCollector(w.client).RetrieveOne(
			ctx,
			obj,
			[]string{extraConfigNamespaceNamePropPath},
			&content,
		)
		if err != nil {
			return err
		}
		namespace, name = namespacedNameFromObjContent(content)
```